### ISSUE TITLE
Failure to build on Python 3.9.1 / Apple arm64

### ISSUE THREAD

**Issue by anentropic** [gevent/gevent#1721]
* gevent version: 20.9.0
* Python version: cPython 3.9.1 downloaded from python.org
* Operating System: macOS 11.1 (on M1 arm64)

### Description:

I have tried two ways to get a python project, which uses gevent, running on a new M1 macbook.

The first way was to use this method https://stackoverflow.com/a/64885034/202168
i.e. clone the Terminal.app and run it under Rosetta, install everything as usual for the project (Python 3.8.6, virtualenv) and in theory it's running as emulated x64 arch.

In that case pip install completed fine but I get an error when I import gevent:
```
Python 3.8.6 (default, Dec 15 2020, 19:24:48)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import gevent
zsh: illegal hardware instruction  ipython
```
(other packages like `psycopg2` import fine, so I don't think it's an ipython problem)

(Activity Monitor reports this ipython `python3.8` process as "intel" architecture, so Rosetta must be active, but clearly was unable to emulate something properly)

Second method I tried was to use the arm64 build of Python 3.9.1, with the installer downloaded from here: https://www.python.org/downloads/release/python-391/

This time I get an error in install phase for gevent:

```
% VENV/bin/pip install --no-binary :all: --no-cache-dir gevent                 
Collecting gevent
  Downloading gevent-20.9.0.tar.gz (5.8 MB)
     |████████████████████████████████| 5.8 MB 2.1 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... error
    ERROR: Command errored out with exit status 1:
     command: /Users/anentropic/Documents/Dev/myproject/VENV/bin/python3 /Users/anentropic/Documents/Dev/myproject/VENV/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel /var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/tmp45tz0spa
         cwd: /private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-install-pffddqig/gevent_6639e99d258f4982b9369a7d01ae5be8
    Complete output (43 lines):
    Traceback (most recent call last):
      File "/Users/anentropic/Documents/Dev/myproject/VENV/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 280, in <module>
        main()
      File "/Users/anentropic/Documents/Dev/myproject/VENV/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 263, in main
        json_out['return_val'] = hook(**hook_input['kwargs'])
      File "/Users/anentropic/Documents/Dev/myproject/VENV/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py", line 133, in prepare_metadata_for_build_wheel
        return hook(metadata_directory, config_settings)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 161, in prepare_metadata_for_build_wheel
        self.run_setup()
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 253, in run_setup
        super(_BuildMetaLegacyBackend,
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/build_meta.py", line 145, in run_setup
        exec(compile(code, __file__, 'exec'), locals())
      File "setup.py", line 471, in <module>
        run_setup(EXT_MODULES)
      File "setup.py", line 338, in run_setup
        setup(
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/__init__.py", line 153, in setup
        return distutils.core.setup(**attrs)
      File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/core.py", line 108, in setup
        _setup_distribution = dist = klass(attrs)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 423, in __init__
        _Distribution.__init__(self, {
      File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/distutils/dist.py", line 292, in __init__
        self.finalize_options()
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 695, in finalize_options
        ep(self)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/setuptools/dist.py", line 702, in _finalize_setup_keywords
        ep.load()(self, ep.name, value)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/cffi/setuptools_ext.py", line 219, in cffi_modules
        add_cffi_module(dist, cffi_module)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/cffi/setuptools_ext.py", line 49, in add_cffi_module
        execfile(build_file_name, mod_vars)
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/cffi/setuptools_ext.py", line 25, in execfile
        exec(code, glob, glob)
      File "src/gevent/libev/_corecffi_build.py", line 31, in <module>
        ffi = FFI()
      File "/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/cffi/api.py", line 48, in __init__
        import _cffi_backend as backend
    ImportError: dlopen(/private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so, 2): Symbol not found: _ffi_prep_closure
      Referenced from: /private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so
      Expected in: flat namespace
     in /private/var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/pip-build-env-jqwff0br/overlay/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so
    ----------------------------------------
ERROR: Command errored out with exit status 1: /Users/anentropic/Documents/Dev/myproject/VENV/bin/python3 /Users/anentropic/Documents/Dev/myproject/VENV/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py prepare_metadata_for_build_wheel /var/folders/w1/_vgkxyln4c7bk8kr29s1y1k00000gn/T/tmp45tz0spa Check the logs for full command output.
```

**Comment by jamadden** [comment]
Thanks for the report!

Looking closely at the end of the report:
```
   File "/lib/python3.9/site-packages/cffi/api.py", line 48, in __init__
        import _cffi_backend as backend
    ImportError: dlopen(//overlay/lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so, 2): Symbol not found: _ffi_prep_closure
      Referenced from: //lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so
      Expected in: flat namespace
     in ///lib/python3.9/site-packages/_cffi_backend.cpython-39-darwin.so
```

We can see that the problem lies in the [CFFI library](https://foss.heptapod.net/pypy/cffi/) dependency, so there's not much I can do about that. (It looks like a linking issue of some sort.) Perhaps they can help on the linked issue tracker?

It is *currently* possible to build gevent without the CFFI dependency, so long as you don't want to use some of its features (notably the libuv backend). That can't be done directly via pip from a released tarball though. You'll need to either download the tarball and unpack it somewhere manually, or clone the repo (and check out the 20.9.0 tag). Next, you'll need to edit the `pyproject.toml` file and comment out the line referring to CFFI. Once that's done, you can proceed by `pip install .` inside that same directory (or `pip wheel .` to get a wheel you can move around and install easily on many machines).

Good luck!

(As an aside, `import gevent -> zsh: illegal hardware instruction  ipython` is interesting and somewhat surprising. The binary wheels are compiled for generic x86_64 architectures, with no specific flags to use the instructions that aren't supported, e.g., the advanced vector instructions. I know they work on relatively old 2011-era hardware. I also know that greenlet is the same. It's curious that rosetta2 isn't handling whatever that is.)

**Comment by jamadden** [comment]
I do find that the CFFI maintainers [specifically state that CFFI doesn't support M1 macs](https://foss.heptapod.net/pypy/cffi/-/issues/464#note_147550):

> I'm going to remove ARM-on-Mac support because it causes troubles (479) on Intel-on-Mac 11. There's additionally 477 to look at, too. If you are motivated to understand and fix these two issues, please go ahead and fix them both. In the meantime, cffi won't support ARM-on-Mac.

**Comment by tyates-indeed** [comment]
cffi has just recently released version 1.15 which comes with wheels for MacOS arm64: https://cffi.readthedocs.io/en/latest/whatsnew.html#v1-15-0

I confirmed that I can successfully run `pip install cffi==1.15` on my M1 Mac.

However, when I try to `pip install gevent` it still tries to build from source which fails. Can someone build a new wheel for gevent that uses the new wheel from cffi version 1.15?

**Comment by tyates-indeed** [comment]
@jamadden now that CFFI has been updated with M1 support, I believe the "status: cantfix" and "status: not gevent" labels should be removed from this issue.

**Comment by tyates-indeed** [comment]
@jamadden is there a chance we can get this merged soon? Intel Macs have been discontinued so this is going to become more and more of a problem.

**Comment by jamadden** [comment]
Is that change included in the upstream repository? (See http://git.savannah.gnu.org/gitweb/?p=config.git;a=tree). I'd happily accept a PR that updated to the latest from upstream if so.

**Comment by tyates-indeed** [comment]
No those changes have not been pushed upstream. https://cffi.readthedocs.io/en/latest/whatsnew.html#v1-15-0 does state:
`MacOS arm64 support (including wheels)`

Could you provide guidance on the best way to proceed here?

**Comment by csalch-indeed** [comment]
It looks like the config.git is already patched upstream. The libev maintainer seems to recommend just updating `config.sub` and config.guess` so I'll update send over a patch with the version packaged in automake 1.16.5. The original libev-4.33 source has an `autogen.sh` script in it that pulls that in for me and it look like c-ares recommend running `autoreconf -fi` though I think that's alot more invasive so I'll just send over a config.sub/guess update unless otherwise requested ..


I'll update my patch here in a bit if everything compiles locally for me.

**Comment by jamadden** [comment]
> It looks like the config.git is already patched upstream. 

Thanks, that's what I was hoping for.

> The libev maintainer seems to recommend just updating `config.sub` and `config.guess` 

As does gevent; we've been in similar situations before and we just wait until the auto* stuff is updated and pull those in. [deps/README.rst](https://github.com/gevent/gevent/blob/master/deps/README.rst):

> Check if 'config.guess' and/or 'config.sub' went backwards in time (the 'timestamp' and 'copyright' dates). If so, revert it (or update from the latest source http://git.savannah.gnu.org/gitweb/?p=config.git;a=tree )

> I'll update my patch here in a bit if everything compiles locally for me.

It would be very helpful if you could also run the gevent test suite. (See here](http://www.gevent.org/development/getting_started.html) for setup and [here](http://www.gevent.org/development/running_tests.html) for running tests; don't worry about running any more than one Python version.) It's nice to have some indication that things work, since the various upstreams may or may not be doing any testing on their own.

**Comment by csalch-indeed** [comment]
I've update https://github.com/gevent/gevent/pull/1828 I'll go through the test suite shortly.

**Comment by doublethefish** [comment]
As noted above the install is trying to incorrectly `/usr/local` intel variant of the lib and not `/opt/local` arm64 variant. This happens same error happens with updated `config.sub` and `config.guess`.

This looks like an `@rpath` issue, but I have run out of time to look into this more. It's probably another part of the configuration scripts outside of the config.sub/guess files.

Here's a few things I did/noticed/found.

1. Run install in verbose and noclean mode `python3 -m pip install -e <path-to-gevent> -v --no-clean`

```
<snip>
  ImportError: dlopen(<snip>/overlay/lib/python3.8/site-packages/_cffi_backend.cpython-38-darwin.so, 0x0002): Library not loaded: @rpath/libffi.7.dylib
    Referenced from: <snip>/overlay/lib/python3.8/site-packages/_cffi_backend.cpython-38-darwin.so
    Reason: tried: '/usr/local/lib/libffi.7.dylib' (no such file), '/usr/lib/libffi.7.dylib' (no such file)
  Preparing metadata (pyproject.toml) ... error
<snip>
```

2. Notice the `@rpath/libffi.7.dylib`
3. Notice it only tried `/usr/local` as a candidate, suggesting a missing or incorrect rpath option - `/usr/local` being the homebrew intel path, not the `/opt/local` silicon homebrew path.
3. Confirm with running `otool -L <snip>/overlay/lib/python3.8/site-packages/_cffi_backend.cpython-38-darwin.so`

```
/private/var/folders/fw/9c4l1c896_d60vtbtwffr8040000gn/T/pip-build-env-75a46i40/overlay/lib/python3.8/site-packages/_cffi_backend.cpython-38-darwin.so:
	@rpath/libffi.7.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
```

4. Note that at the bottom of _coercffi_build.py there is the following text, which suggests a possible next step (look at that portion of the build).

```
    # On macOS to make the non-embedded case work correctly, against
    # our local copy of libev:
    #
    # 1) configure and make libev
    # 2) CPPFLAGS=-Ideps/libev/ LDFLAGS=-Ldeps/libev/.libs GEVENTSETUP_EMBED_LIBEV=0 \
    #     python setup.py build_ext -i
    # 3) export DYLD_LIBRARY_PATH=`pwd`/deps/libev/.libs
    #
    # The DYLD_LIBRARY_PATH is because the linker hard-codes
    # /usr/local/lib/libev.4.dylib in the corecffi.so dylib, because
    # that's the "install name" of the libev dylib that was built.
    # Adding a -rpath to the LDFLAGS doesn't change things.
    # This can be fixed with `install_name_tool`:
    #
    # 3) install_name_tool -change /usr/local/lib/libev.4.dylib \
    #    `pwd`/deps/libev/.libs/libev.4.dylib \
    #     src/gevent/libev/_corecffi.abi3.so
    ffi.compile(verbose=True)
```
5. Setting custom `LDFLAGS` (taken from `brew info libffi`) doesn't seem to dix the issue e.g. `LDFLAGS="-L/opt/homebrew/opt/libffi/lib" python3 -m pip install -e <path-to-gevent> -v --no-clean`.

6. Changing libev's configure to inherit rpaths didn't work, but I am not sure what paths it had in the first place to inherit.

```diff
+++ b/deps/libev/configure
@@ -8917,7 +8917,7 @@ $as_echo_n "checking whether the $compiler linker ($LD) supports shared librarie
   hardcode_libdir_separator=
   hardcode_minus_L=no
   hardcode_shlibpath_var=unsupported
-  inherit_rpath=no
+  inherit_rpath=yes
```

Although I am not sure, I think this issue is a gevent specific issue, and its use of cffi. A quick hack might be to just add the homebrew libpath as an rapth or respect/passthru  LDFLAGS

**Comment by tyates-indeed** [comment]
Okay I believe the last step here now that https://github.com/gevent/gevent/issues/1839 is closed is to publish a new version of gevent with wheels for arm64. Is that correct?

**Comment by artinnok** [comment]
Hey guys!

As I understand #1839 was included into latest `gevent`  [release](http://www.gevent.org/changelog.html#bugfixes).
Any updates with this issue?

**Comment by doublethefish** [comment]
Still the same.

Looking at it with fresh eyes, the issue is upstream in [libffi's](https://github.com/libffi/libffi) python module, which just needs to add the relevant rpath to the module.

If I get time/chance, I'll create a repo case and open an issue there.

**Comment by doublethefish** [comment]
tldr; works with a clean env.

It looks like my env was dirty and still had the old backend dylib, so it is working, with caveats.

Luckily I still had it up; I had run `python3 -m pip install .` (notice the missing `--no-cache-dir`*) and my logs show it was using a version of `_cffi_backend.cpython-38-darwin.so` that referenced `@rpath/libffi.7.dylib`, I now assume because `_cffi_backend.cpython-38-darwin.so` was stale. 

Am I correct in thinking that `_cffi_backend.cpython-38-darwin.so` is actually built by `src/gevent/libev/_corecffi_build.py`?  - and is not a cffi binary, but [built by it](https://cffi.readthedocs.io/en/latest/cdef.html) at install-time.

I would quite like to understand how I had the dirty state and why setuptools didn't update rebuild the backend `.so` at `./.pyenv/lib/python3.8/site-packages/_cffi_backend.cpython-38-darwin.so`. I assume there's a missing dependency chain.

\* I don't actually think `--no-cache-dir` helps in this case.

**Comment by tahesse** [comment]
I had issues installing `gevent` in my python 3.8 virtualenv. I was able to install it to my global site-packages however.

**Comment by Maleware** [comment]
Hey,

back at this topic again.. Trying to build gevent from scratch for arm64 as I would like to get airflow to build for arm64. If I try to do so:
`In file included from src/gevent/_greenlet_primitives.c:1207:
      /private/var/folders/zs/mlhz8w0d6vsb6_95tkgh_v7c0000gn/T/pip-install-wi26uynz/gevent_d30430066b2642e481079ebe09a733ce/deps/greenlet/greenlet.h:42:5: error: unknown type name 'CFrame'
          CFrame* cframe;
          ^
      1 error generated.
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]`
      
This looks to me like a problem in the gevent code for arm64, maybe because this is not an atomic type?

Hope for help and greetz

**Comment by jamadden** [comment]
> `deps/greenlet/greenlet.h:42:5: error: unknown type name 'CFrame' CFrame* cframe`
> This looks to me like a problem in the gevent code for arm64, maybe because this is not an atomic type?

No, it's definitely not that. greenlet and gevent are regularly built for macOS Arm64 and Linux aarch64. Most likely you're trying to install incompatible versions, i.e., you have a mismatch between greenlet and your version of Python.

Closing this because gevent and greenlet have been building on macOS arm64 for well over a year.

**Comment by Maleware** [comment]
I highly doubt this since the configuration builds for AMD. However, feel free to test this by building airflow from scratch on arm64. 

If you use gevent 23.9.1 and greenlet 3.0.2 gevent can't be build.

However, it works with gevent 22.10.2

**Comment by jamadden** [comment]
> If you use gevent 23.9.1 and greenlet 3.0.2 gevent can't be build.

Nope, you are 100% wrong. Want proof? Both of those project versions publish binary wheels for that configuration on PyPI! You have an issue in your environment.

**Comment by Maleware** [comment]
Yeah I know. I was wondering myself. But... it's hard to imagine that this is about the env because we have 100% control over it: https://github.com/stackabletech/docker-images/blob/main/airflow/Dockerfile

This is tested an released on amd and... just works. Can you spot the mistake?
